File: C++/src/Compiler.cpp
#include "Compiler.hpp"
#include "Error.hpp"
#include "Lexer.hpp"

#include <stack>

namespace bfc
{
    struct BracePosition
    {
        bfl::SourcePosition SrcPosition;
        std::size_t CodePosition;

        BracePosition(bfl::SourcePosition src, std::size_t code) :
            SrcPosition(src), CodePosition(code) {}
    };

    static void ParseProgram(std::vector<ByteCode>& out);
    static void ParseAddByte(std::vector<ByteCode>& out);
    static void ParseSubByte(std::vector<ByteCode>& out);
    static void ParseAddPtr(std::vector<ByteCode>& out);
    static void ParseSubPtr(std::vector<ByteCode>& out);
    static void ParseWrite(std::vector<ByteCode>& out);
    static void ParseRead(std::vector<ByteCode>& out);
    static void ParseConditional(std::vector<ByteCode>& out);
    static void ParseChain(Token token, OpCode op, std::vector<ByteCode>& out);

    static Token s_CurrentToken = Token::NONE;
    static std::size_t s_CurrentLine = 0;

    void Init(std::string_view filepath)
    {
        LexerInit(filepath);
        bfl::SetProgramName(filepath);
    }

    std::vector<ByteCode> Compile()
    {
        std::vector<ByteCode> out;
        ParseProgram(out);

        return out;
    }

    static void ParseProgram(std::vector<ByteCode>& out)
    {
        s_CurrentToken = GetToken();
        while (s_CurrentToken != Token::END_OF_FILE)
        {
            switch (s_CurrentToken)
            {
                case Token::PLUS:
                    ParseAddByte(out);
                    break;
                case Token::MINUS:
                    ParseSubByte(out);
                    break;
                case Token::ARROW_LEFT:
                    ParseSubPtr(out);
                    break;
                case Token::ARROW_RIGHT:
                    ParseAddPtr(out);
                    break;
                case Token::DOT:
                    ParseWrite(out);
                    break;
                case Token::COMMA:
                    ParseRead(out);
                    break;
                case Token::BRACE_LEFT:
                    ParseConditional(out);
                    break;
                default:
                    bfl::LogErrorPosition("invalid token: %d", static_cast<int32_t>(s_CurrentToken));
                    break;
            }
        }

        out.emplace_back(OpCode::END);
    }

    static void ParseAddByte(std::vector<ByteCode>& out)
    {
        ParseChain(Token::PLUS, OpCode::ADDB, out);
    }

    static void ParseSubByte(std::vector<ByteCode>& out)
    {
        ParseChain(Token::MINUS, OpCode::SUBB, out);
    }

    static void ParseAddPtr(std::vector<ByteCode>& out)
    {
        ParseChain(Token::ARROW_RIGHT, OpCode::ADDP, out);
    }

    static void ParseSubPtr(std::vector<ByteCode>& out)
    {
        ParseChain(Token::ARROW_LEFT, OpCode::SUBP, out);
    }

    static void ParseWrite(std::vector<ByteCode>& out)
    {
        out.emplace_back(OpCode::WRITE);
        s_CurrentToken = GetToken();

        s_CurrentLine++;
    }

    static void ParseRead(std::vector<ByteCode>& out)
    {
        out.emplace_back(OpCode::READ);
        s_CurrentToken = GetToken();

        s_CurrentLine++;
    }

    static void ParseConditional(std::vector<ByteCode>& out)
    {
        std::stack<BracePosition, std::vector<BracePosition>> braces;

        braces.emplace(bfl::g_Position, s_CurrentLine++);
        out.emplace_back(OpCode::JZ);

        s_CurrentToken = GetToken();
        while (!braces.empty())
        {
            if (s_CurrentToken == Token::END_OF_FILE)
            {
                bfl::g_Position = braces.top().SrcPosition;
                bfl::LogErrorPosition("no matching ']'");
            }

            switch (s_CurrentToken)
            {
                case Token::PLUS:
                    ParseAddByte(out);
                    break;
                case Token::MINUS:
                    ParseSubByte(out);
                    break;
                case Token::ARROW_LEFT:
                    ParseSubPtr(out);
                    break;
                case Token::ARROW_RIGHT:
                    ParseAddPtr(out);
                    break;
                case Token::DOT:
                    ParseWrite(out);
                    break;
                case Token::COMMA:
                    ParseRead(out);
                    break;
                case Token::BRACE_LEFT:
                {
                    braces.emplace(bfl::g_Position, s_CurrentLine++);
                    out.emplace_back(OpCode::JZ);

                    s_CurrentToken = GetToken();
                } break;
                case Token::BRACE_RIGHT:
                {
                    const std::size_t open = braces.top().CodePosition;

                    out[open].Line = ++s_CurrentLine;
                    out.emplace_back(OpCode::JMP, open);

                    s_CurrentToken = GetToken();
                    braces.pop();
                } break;
                default:
                    bfl::LogErrorPosition("invalid token: %d", static_cast<int32_t>(s_CurrentToken));
                    break;
            }
        }
    }

    static inline void ParseChain(Token token, OpCode op, std::vector<ByteCode>& out)
    {
        ByteCode& code = out.emplace_back(op);

        while (s_CurrentToken == token)
        {
            code.PointerOffset++;
            s_CurrentToken = GetToken();
        }

        s_CurrentLine++;
    }
}

File: Java/src/Compiler.java

import java.util.ArrayList;
import java.util.Stack;

public class Compiler {
    
    private Lexer lexer;
    private int currentLine;
    private Token currentToken;

    public Compiler(String filePath) {
        this.lexer = new Lexer(filePath);
        this.currentLine = 0;
        this.currentToken = Token.NONE;
    }

    public ArrayList<ByteCode> compile() {
        final ArrayList<ByteCode> code = new ArrayList<ByteCode>();

        currentToken = lexer.nextToken();
        while (currentToken != Token.END_OF_FILE) {
            switch (currentToken) {
                case PLUS: parseAddByte(code); break;
                case MINUS: parseSubByte(code); break;
                case ARROW_LEFT: parseSubPtr(code); break;
                case ARROW_RIGHT: parseAddPtr(code); break;
                case DOT: parseWrite(code); break;
                case COMMA: parseRead(code); break;
                case BRACKET_LEFT: parseConditional(code); break;
                default: {
                    System.err.println("bfvm: error: unexpected token");
                    System.exit(1);
                } break;
            }
        }

        code.add(new ByteCode(OpCode.END, 0));
        lexer.close();
        
        return code;
    }

    private void parseAddByte(ArrayList<ByteCode> code) {
        int val = 0;

        while (currentToken == Token.PLUS) {
            val++;
            currentToken = lexer.nextToken();
        }

        code.add(new ByteCode(OpCode.ADDB, val));
        currentLine++;
    }

    private void parseSubByte(ArrayList<ByteCode> code) {
        int val = 0;

        while (currentToken == Token.MINUS) {
            val++;
            currentToken = lexer.nextToken();
        }

        code.add(new ByteCode(OpCode.SUBB, val));
        currentLine++;
    }

    private void parseAddPtr(ArrayList<ByteCode> code) {
        int offset = 0;

        while (currentToken == Token.ARROW_RIGHT) {
            offset++;
            currentToken = lexer.nextToken();
        }

        code.add(new ByteCode(OpCode.ADDP, offset));
        currentLine++;
    }

    private void parseSubPtr(ArrayList<ByteCode> code) {
        int offset = 0;

        while (currentToken == Token.ARROW_LEFT) {
            offset++;
            currentToken = lexer.nextToken();
        }

        code.add(new ByteCode(OpCode.SUBP, offset));
        currentLine++;
    }

    private void parseWrite(ArrayList<ByteCode> code) {
        code.add(new ByteCode(OpCode.WRITE, 0));
        currentToken = lexer.nextToken();

        currentLine++;
    }

    private void parseRead(ArrayList<ByteCode> code) {
        code.add(new ByteCode(OpCode.READ, 0));
        currentToken = lexer.nextToken();

        currentLine++;
    }

    private void parseConditional(ArrayList<ByteCode> code) {
        final Stack<Integer> braces = new Stack<Integer>();

        braces.push(currentLine++);

        code.add(new ByteCode(OpCode.JZ, 0));
        currentToken = lexer.nextToken();
        while (!braces.empty()) {
            if (currentToken == Token.END_OF_FILE) {
                System.err.println("bfvm: error: unbalanced braces");
                System.exit(1);
            }

            switch (currentToken) {
                case PLUS: parseAddByte(code); break;
                case MINUS: parseSubByte(code); break;
                case ARROW_LEFT: parseSubPtr(code); break;
                case ARROW_RIGHT: parseAddPtr(code); break;
                case DOT: parseWrite(code); break;
                case COMMA: parseRead(code); break;
                case BRACKET_LEFT: {
                    braces.push(currentLine++);

                    code.add(new ByteCode(OpCode.JZ, 0));
                    currentToken = lexer.nextToken();
                } break;
                case BRACKET_RIGHT: {
                    final int open = braces.pop();
                    final ByteCode openingBrace = code.get(open);
                    
                    openingBrace.operand = ++currentLine;
                    code.add(new ByteCode(OpCode.JMP, open));
                    
                    currentToken = lexer.nextToken();
                } break;
                default: {
                    System.err.println("bfvm: error: unexpected token");
                    System.exit(1);
                } break;
            }
        }
    }
}
